### PR TITLE
enable injecting development versions

### PR DESCRIPTION
**Pull request by RsrchBoy**
This pokes Parse::LocalDistribution in the correct way to get it indexing
development versions (e.g. 0.01_01).  It also adds --allow-dev (though doesn't
document it at the moment).

This is at least a partial fix to issue #22.

### CODE REVIEW

**Comment by oalders** [comment]
Great! Is there a chance you could also add a test for this?

**Comment by RsrchBoy** [comment]
On 04/03/2015 06:19 AM, Olaf Alders wrote:
>
> Great! Is there a chance you could also add a test for this?
>

There is :)  But I have the kidlets this weekend; so likely not until
mid-next week at the earliest.

**Comment by exodist** [comment]
+1

**Comment by oalders** [comment]
@RsrchBoy "time with kids" > "writing unit tests".  :)  I may be able to find some time for this at the QA Hackathon this week.

**Comment by RsrchBoy** [comment]
@oalders Thanks!  Normally, no problem (and I'm kinda embarrassed for submitting it without any), but the valuation you stated controls =)

**Comment by oalders** [comment]
@tokuhirom Can I get your thoughts on this?  It appears to me that OrePAN2 is not blocking the indexing of all dev releases by default.  Consider the following cases:

`orepan2-inject --text --simple http://cpan.metacpan.org/authors/id/T/TS/TSIBLEY/App-RecordStream-4.0.8-TRIAL.tar.gz foo`

That appears to index everything in the dist. This module has no "provides" metadata, but the versions also have no underscores.  Clearly, from the name, it's a dev release.

`orepan2-inject --text --simple https://cpan.metacpan.org/authors/id/R/RI/RIBASUSHI/DBIx-Class-0.082899_14.tar.gz foo`

This indexes everything except for DBIx::Class itself.  It also indexes the following modules which are bundled into the dist but are really part of SQL::Translator:

```
SQL::Translator::Parser::DBIx::Class 1.10                   D/DU/DUMMY/DBIx-Class-0.082899_14.tar.gz
SQL::Translator::Producer::DBIx::Class::File 0.1                    D/DU/DUMMY/DBIx-Class-0.082899_14.tar.gz
```

Those are just two quick examples to show that the current behaviour is a bit confusing.  I think we could either:

1. Work really hard to exclude dev releases properly
2. Just allow dev releases by default and assume that anyone who really cares about this will either provide a patch or "do the right thing" before blindly passing tarballs for injection and/or indexing

I'm inclined to go with the second option but I could be convinced to go with the first with a compelling argument.  What do you think?

**Comment by RsrchBoy** [comment]
...from the peanut gallery: second option ++.  There are valid reasons to include dev releases, and it's very surprising that dev releases would be excluded from indexing even when explicitly specified/provided.

**Comment by RsrchBoy** [comment]
(but then, I'm biased ;))

**Comment by tokuhirom** [comment]
second option ++

**Comment by oalders** [comment]
OK, let's do that.  @RsrchBoy do you want to amend this to enable dev releases across the board with no command line options required?  We can't really disable it reliably anyway, AFAICT.  Don't think that will require any additional tests, to be honest.

**Comment by RsrchBoy** [comment]
Awesome.  I'll take a whack at it in a little bit :)

File: lib/OrePAN2/Indexer.pm [hunk 1]
```diff
@@ -61,7 +62,7 @@ sub add_index {
     return if $self->_maybe_index_from_metacpan( $index, $archive_file );
 
     my $archive = Archive::Extract->new( archive => $archive_file );
-    my $tmpdir = tempdir( CLEANUP => 1 );
+    my $tmpdir = tempdir('orepan2.XXXXXX', TMPDIR => 1, CLEANUP => 1 );
```
Comment (oalders) [inline lib/OrePAN2/Indexer.pm]: What's the reason for this change?

**Comment by oalders** [comment]
I've reworked this a bit and merged into master.  Thanks for all of your work on this @RsrchBoy. :)  Will release to CPAN today.
